### PR TITLE
[HINTS]Add implementation for hints on functions compute_slope and ec_double (cairo_secp/ec.cairo )

### PR DESCRIPTION
**Pull request by pefontana**
# [HINTS]Add implementation for hints on functions compute_slope and ec_double (cairo_secp/ec.cairo )

## Description

[cairo_secp/ec.cairo](https://github.com/starkware-libs/cairo-lang/blob/167b28bcd940fd25ea3816204fa882a0b0a49603/src/starkware/cairo/common/cairo_secp/ec.cairo#L93)

* Hints implementation:

```
compute_slope:
%{
    from starkware.cairo.common.cairo_secp.secp_utils import SECP_P, pack
    from starkware.python.math_utils import line_slope

    # Compute the slope.
    x0 = pack(ids.point0.x, PRIME)
    y0 = pack(ids.point0.y, PRIME)
    x1 = pack(ids.point1.x, PRIME)
    y1 = pack(ids.point1.y, PRIME)
    value = slope = line_slope(point1=(x0, y0), point2=(x1, y1), p=SECP_P)
%}
```

```
ec_double_assign_new_x:
%{
    from starkware.cairo.common.cairo_secp.secp_utils import SECP_P, pack

    slope = pack(ids.slope, PRIME)
    x = pack(ids.point.x, PRIME)
    y = pack(ids.point.y, PRIME)

    value = new_x = (pow(slope, 2, SECP_P) - 2 * x) % SECP_P
%}
```

```
ec_double_assign_new_y:
%{ value = new_y = (slope * (x - new_x) - y) % SECP_P %}
```

## Checklist
- [ ] Linked to Github Issue
- [x] Unit tests added
- [x] Integration tests added.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.

### CODE REVIEW

File: src/vm/hints/secp/ec_utils.rs [hunk 1]
```diff
@@ -91,6 +91,161 @@ pub fn compute_doubling_slope(
...
+        get_integer_from_relocatable_plus_offset(&point_reloc, 3, vm)?,
+        get_integer_from_relocatable_plus_offset(&point_reloc, 4, vm)?,
+        get_integer_from_relocatable_plus_offset(&point_reloc, 5, vm)?,
+    );
+
+    let slope = pack(slope_d0, slope_d1, slope_d2, &vm.prime);
+    let x = pack(x_d0, x_d1, x_d2, &vm.prime);
+    let y = pack(y_d0, y_d1, y_d2, &vm.prime);
+
+    let value = (slope.pow(2).mod_floor(&SECP_P) - (&x * 2_usize)).mod_floor(&SECP_P);
```
Comment (Oppen) [inline src/vm/hints/secp/ec_utils.rs:200]: Modulus is transitive to substraction, shift by 1 is cheaper than multiplication by 2:
```suggestion
    let value = (slope.pow(2) - (&x << 1_usize)).mod_floor(&SECP_P);
```

File: src/vm/hints/secp/ec_utils.rs [hunk 2]
```diff
@@ -247,4 +400,335 @@ mod tests {
...
+        let hint_code = "from starkware.cairo.common.cairo_secp.secp_utils import SECP_P, pack\nfrom starkware.python.math_utils import line_slope\n\n# Compute the slope.\nx0 = pack(ids.point0.x, PRIME)\ny0 = pack(ids.point0.y, PRIME)\nx1 = pack(ids.point1.x, PRIME)\ny1 = pack(ids.point1.y, PRIME)\nvalue = slope = line_slope(point1=(x0, y0), point2=(x1, y1), p=SECP_P)";
+        let mut vm = VirtualMachine::new(
+            VM_PRIME.clone(),
+            vec![(
+                "range_check".to_string(),
+                Box::new(RangeCheckBuiltinRunner::new(true, bigint!(8), 8)),
+            )],
+            false,
+            &HINT_EXECUTOR,
+        );
```
Comment (Oppen) [inline src/vm/hints/secp/ec_utils.rs:415]: There's a newly added macro for this, `vm_with_range_check!` IIRC.

File: src/vm/hints/secp/ec_utils.rs [hunk 3]
```diff
@@ -247,4 +400,335 @@ mod tests {
...
+                    dereference: false,
+                    inner_dereference: false,
+                    immediate: None,
+                    ap_tracking_data: Some(ApTracking {
+                        group: 1,
+                        offset: 0,
+                    }),
+                },
+            ),
+        ]);
```
Comment (Oppen) [inline src/vm/hints/secp/ec_utils.rs:471]: And another one for references.

**Review by Oppen — CHANGES_REQUESTED** [review]

File: src/vm/hints/secp/ec_utils.rs (hunk 1 above)
Comment (pefontana) [reply src/vm/hints/secp/ec_utils.rs]: I knew you would find a better way to do this, but I could figure out a better one :P 
Done

File: src/vm/hints/secp/ec_utils.rs (hunk 2 above)
Comment (pefontana) [reply src/vm/hints/secp/ec_utils.rs]: Nice, didnt see it.
Done

File: src/vm/hints/secp/ec_utils.rs (hunk 3 above)
Comment (pefontana) [reply src/vm/hints/secp/ec_utils.rs:448]: Yes, I saw the references! macro, but in these cases where the offset1 is not auto-increment by 1, is not useful

```
for i in 0..$num {
    references.insert(i, HintReference::new_simple((i as i32 - $num)));
```
```
pub fn new_simple(offset1: i32) -> Self {
    HintReference {
        register: Register::FP,
        offset1,
        offset2: 0,
```

Comment (Oppen) [reply src/vm/hints/secp/ec_utils.rs:448]: Oh. We need a smarter one then.

**Review by Oppen — APPROVED** [review]

File: src/vm/hints/secp/ec_utils.rs (hunk 3 above)
Comment (pefontana) [reply src/vm/hints/secp/ec_utils.rs:448]: I will implement one in the following PR

**Pull request merged**